src/v/cluster/rm_stm.cc

```diff
@@ -2515,8 +2516,10 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
               rm_stm::clear_type::idempotent_pids);
         }
 
-        seq_it->second.entry.last_write_timestamp = bid.first_timestamp.value();
-        _oldest_session = std::min(_oldest_session, bid.first_timestamp);
+        seq_it->second.entry.last_write_timestamp = bid.max_timestamp.value();
```

Comment:
We can't change this without changing the dependencies of `last_write_timestamp` & `_oldest_session` otherwise it breaks implicit invariant.

For example, `_oldest_session = std::min(_oldest_session, bid.first_timestamp);` assumes that first_timestamp is the min timestamp of the batch and uses the fact that `min(a,min(b,c)) == min(min(a,b),c)` (associativity of min). If we replace `first_timestamp` with `max_timestamp` then  `_oldest_session` loses its current meaning.

Reply:
The oldest session is only used for `compact_snapshot` the same rule applies for the `last_write_timestamp`. (the `last_write_timestamp` i didn't want to change in this PR but it seems off that it is sometimes updated from wall clock and sometimes using batch header) Both of the fields are used to cleanup producer information. Previously it might be the when using `batch.first_timestamp` as a cleanup boundry that we evicted producer to early  as it might contain records which are not yet pass the deadline. 

I think this change doesn't change anything for the dependencies of `last_write_timestamp` and `_oldest_session`

Reply:
I was checking `last_write_timestamp` and it doesn't seem to be updated correctly, it is initialized with the first time stamp on the batch and later updated to wall clock time, looks off .. but to me it seems like the associativity of min also applies similarly to max_timestamp.

Reply:
I've checked too, it looks like nothing got broken